src/grt/src/GlobalRouter.cpp

```diff
@@ -2269,6 +2346,39 @@ bool GlobalRouter::segmentIsLine(const GSegment& segment)
   return (dimensionality == 1);
 }
 
+bool GlobalRouter::segmentCoversPin(const GSegment& segment, const Pin& pin)
+{
+  auto [min_x, max_x] = std::minmax(segment.init_x, segment.final_x);
+  auto [min_y, max_y] = std::minmax(segment.init_y, segment.final_y);
+  auto [min_layer, max_layer]
+      = std::minmax(segment.init_layer, segment.final_layer);
+  return (pin.getOnGridPosition().getX() >= min_x
+          && pin.getOnGridPosition().getX() <= max_x
+          // Pin is vertically covered by the segment
+          && pin.getOnGridPosition().getY() >= min_y
+          && pin.getOnGridPosition().getY() <= max_y
+          // Pin and segment share a layer
+          && pin.getConnectionLayer() >= min_layer
+          && pin.getConnectionLayer() <= max_layer);
+}
+
+// Builds the Net Graph in O(N²)
+std::vector<std::vector<int>> GlobalRouter::buildNetGraph(odb::dbNet* net)
+{
+  int total_segments = routes_[net].size();
+  std::vector<std::vector<int>> graph(total_segments, std::vector<int>());
+  for (int i = 0; i < total_segments; i++) {
+    for (int j = i - 1; j >= 0; j--) {
+      if (!segmentsConnect(routes_[net][i], routes_[net][j])) {
```

Comment:
You can create a const& for routes_[net], avoiding multiple access to the routes_ map.